pkg/controller/cloud-controller-cap/capability.go

```diff
@@ -0,0 +1,190 @@
+package cloudcontrollercap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	configv1 "github.com/openshift/api/config/v1"
+	openshiftconfigclientset "github.com/openshift/client-go/config/clientset/versioned"
+	configscheme "github.com/openshift/client-go/config/clientset/versioned/scheme"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	kscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+const installConfigFilename = "cluster-config.yaml"
+
+const defaultManifestsDir = "/etc/mcc/bootstrap"
+
+type config struct {
+	manifestDir     string
+	configClientSet openshiftconfigclientset.Interface
+	restConfig      *rest.Config
+}
+
+type Option func(c *config)
+
+func WithManifestDir(dir string) Option {
+	return func(c *config) {
+		c.manifestDir = dir
+	}
+}
+
+func WithConfigClientSet(cl openshiftconfigclientset.Interface) Option {
+	return func(c *config) {
+		c.configClientSet = cl
+	}
+}
+
+func WithRestConfig(cfg *rest.Config) Option {
+	return func(c *config) {
+		c.restConfig = cfg
+	}
+}
+
+// IsCloudControllerCapDisabled checks if CloudController capability disabled
+// through querying kubernetes api for clusterversion or reading capabilities
+// from the install config
+func IsCloudControllerCapDisabled(opts ...Option) (bool, error) {
+	conf := config{
+		manifestDir: defaultManifestsDir,
+	}
+	for _, opt := range opts {
+		opt(&conf)
+	}
+
+	// check if running in hypershift and return true because
+	// hypershift does not support capabilities
+	external, err := isTopologyExternal(conf.manifestDir)
+	if err != nil {
+		klog.Infoln("failed to check if topology is external", err)
+	} else if external {
+		return false, nil
+	}
+
+	disabled, err := checkClusterVersionForCCMODisabled(conf)
+	if err != nil {
+		klog.Errorln("failed to check if CloudController disabled via clusterversion, trying install config", err)
+	} else {
+		return disabled, err
+	}
+
+	disabled, err = checkInstallConfigForCCMODisabled(conf)
```

Comment:
The cluster version should be the source of truth, why is this required? If the clusterversion checks fail I would not expect a fallback